lib/babe/slot.go

```diff
@@ -0,0 +1,55 @@
+package babe
+
+import (
+	"time"
+)
+
+// timeUntilNextSlotInNanos calculates, based on the current system time, the remainng
+// time to the next slot
+func timeUntilNextSlotInMilli(slotDuration time.Duration) time.Duration {
+	now := time.Now().UnixNano()
+	slotDurationInMilli := slotDuration.Nanoseconds()
+
+	nextSlot := (now + slotDurationInMilli) / slotDurationInMilli
+
+	remaining := nextSlot*slotDurationInMilli - now
+	return time.Duration(remaining)
+}
+
+type slotHandler struct {
+	slotDuration time.Duration
+	lastSlot     *Slot
+}
+
+func newSlotHandler(slotDuration time.Duration) *slotHandler {
+	return &slotHandler{
+		slotDuration: slotDuration,
+	}
+}
+
+func (s *slotHandler) waitForNextSlot() Slot {
```

Comment:
perhaps helpful to link the substrate code here? Don't have too, just a possible suggestion if you think it could be helpful. At first I linked above function but I think if a link is anywhere this func is most appropriate